File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> temp;
        recurse(0, candidates, temp, ans, target);
        return ans;
    }

    void recurse(int index, vector<int>&candidates, vector<int>&temp, vector<vector<int>>&ans, int target) {
        if(index>=candidates.size() || target<=0 || candidates[index]>target) {
            if(target==0) ans.push_back(temp);
            return;
        }
        temp.push_back(candidates[index]);
        recurse(index, candidates, temp, ans, target-candidates[index]);
        temp.pop_back();
        recurse(index+1, candidates, temp, ans, target);
    }

};

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let ans = [], temp = [];
    recurse(0,0);
    return ans;
    
    function recurse(index, sum) {
        if(index >= candidates.length || sum>=target) {
            if(sum===target) ans.push(temp.slice()); 
            return;
        }
        temp.push(candidates[index]);
        recurse(index,sum+candidates[index]);
        temp.pop();
        recurse(index+1,sum);
    }
};